### PR TITLE
audio: Add Crossover Filter component

### PR DESCRIPTION
**Pull request by johnylin76**
Note: this PR contains the upstreamed commits from PR: https://github.com/thesofproject/sof/pull/2802

This patch adds a Crossover Filter component to SOF. A crossover can be used to split an input signal into different frequency bands. It's a one input, multiple output component. The number of outputs supported are 2,3 and 4. Supports s16_le, s24_le and s32_le.

To maximize code reuse, the biquad processing function was separated from iir_df2t(). This allows to process the biquads more freely. The motivation behind this change was to reuse the coefficients of the LR4 biquads. An LR4 filter is the basic building block of a crossover, and it is made of two biquads in series with same coefficients. Therefore to save memory, we can store one copy of each set of coeffcients, and pass those to the biquad processing function.

The tool to generate the coefficients are found under tools/tune/crossover, and you just need to run:

`octave example_crossover.m`
If you want tweak the parameters (number of outputs, frequency cutoffs or sink assignments) just modify example_crossover.m.

### CODE REVIEW

File: tools/topology/sof/pipe-crossover-playback.m4
Comment (singalsu) [inline tools/topology/sof/pipe-crossover-playback.m4:26]: It would be good to undefine these in the end of this macro to make sure there's every time a define for them before they are used.

File: src/audio/crossover/crossover.c
Comment (singalsu) [inline src/audio/crossover/crossover.c:181]: Please merge this to previous commit for crossover.c.

File: src/audio/crossover/crossover.c
Comment (singalsu) [inline src/audio/crossover/crossover.c:40]: Also merge this to previous commit.

**Comment by cujomalainey** [comment]
Thanks for the review @singalsu, I agree we should squash fixes into their original commit.

**Comment by johnylin76** [comment]
Thanks for review, commits are updated.

**Comment by singalsu** [comment]
SOFCI TEST

**Comment by cujomalainey** [comment]
@singalsu we are working in #3270 to remove the inlining of the iir function, that way we can use the function as is without any changes and also select the SIMD variants easier, need to resolve that first before landing this

**Comment by johnylin76** [comment]
@singalsu @cujomalainey I have updated this PR to have iir_df2t functions in src/math. The previous inline implementation by Seb is removed.

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:106]: -EINVAL

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:129]: C style comments. (except for SPDX header).

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:77]: inline ?

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:51]: inline ?

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:63]: inline ?

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:42]: inline ?

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:400]: C style

File: src/audio/crossover/crossover.c
Comment (lgirdwood) [inline src/audio/crossover/crossover.c:421]: ditto

File: src/audio/crossover/crossover_generic.c
Comment (lgirdwood) [inline src/audio/crossover/crossover_generic.c:22]: inline ? plus a few below can be inline too

File: src/audio/crossover/crossover_generic.c
Comment (lgirdwood) [inline src/audio/crossover/crossover_generic.c:104]: Can you do sinks on the outer loop, that way it's checked less frequently and will be a faster loop.

File: src/audio/crossover/crossover_generic.c
Comment (lgirdwood) [inline src/audio/crossover/crossover_generic.c:127]: ditto.

File: src/audio/crossover/crossover_generic.c
Comment (lgirdwood) [inline src/audio/crossover/crossover_generic.c:160]: ditto, but not sure if sinks can be outer loop here ?

File: src/include/ipc/topology.h
Comment (lgirdwood) [inline src/include/ipc/topology.h:50]: SOF_COMP_PROCESS should be used here. @keyonjie does your UUID PRs ass this type ? @plbossart fyi

File: src/audio/crossover/crossover_generic.c
Comment (cujomalainey) [reply src/audio/crossover/crossover_generic.c:104]: @lgirdwood I am not sure what you are seeing here? This way we only need to read once and then write n times. If we flipped this it would be read each sample n times and write n times.

File: src/audio/crossover/crossover_generic.c
Comment (cujomalainey) [reply src/audio/crossover/crossover_generic.c:127]: Same comment as above, not sure what the optimization would be

File: src/audio/crossover/crossover_generic.c
Comment (johnylin76) [reply src/audio/crossover/crossover_generic.c:104]: I think @cujomalainey is right. If we do sinks on the outer loop, we will need to call audio_stream_read_frag_s16() to read x per sink.

Comment (lgirdwood) [reply src/audio/crossover/crossover_generic.c:105]: ok, so I guess you always want to increment x regardless of sink. Makes sense.

File: tools/topology/m4/crossover.m4
Comment (keyonjie) [inline tools/topology/m4/crossover.m4:29]: don't need this token if you add UUID for it, e.g. refer to detect.m4

File: tools/tune/crossover/crossover_build_blob.m
Comment (cujomalainey) [inline tools/tune/crossover/crossover_build_blob.m:1]: have you tested setting the config recently? I am worried the blobs don't align still

File: src/audio/CMakeLists.txt
Comment (cujomalainey) [inline src/audio/CMakeLists.txt:122]: this method of moving into math still worries me because won't we need to move the crossover into the math sections as well once we start building the multiband DRC?

Comment (cujomalainey) [reply src/audio/CMakeLists.txt:122]: I am noticing that all the libraries are linked back to SOF but not cross-linked, I wonder if we need need to add a case for cross linking libs

Comment (lgirdwood) [reply src/audio/CMakeLists.txt:122]: @singalsu can you give us your 2c here on the maths library, which functions still need to go in it. Do you have a rough plan of how it should look ? This should mostly be code moving, or is there some coupling that prevents a code move today ?

File: src/include/ipc/topology.h
Comment (johnylin76) [reply src/include/ipc/topology.h:50]: SOF_COMP_PROCESS doesn't seem to exist in tip-of-tree code. Maybe it is still un-merged?

Comment (keyonjie) [reply src/include/ipc/topology.h:50]: No need to add a new comp type enum, just keep this part intact should work to me. @johnylin76

File: tools/tune/crossover/crossover_build_blob.m
Comment (johnylin76) [reply tools/tune/crossover/crossover_build_blob.m:1]: Maybe you're right. After checking the waveform of 2-way and 4-way modified config, the blobs doesn't seem to align still.
(The output waveform looks weird...)
However I still don't get why the alignment is changed? Because I suppose src/audio/crossover/crossover.c parses coeff blob into iir_state_df2t, which is still the input of current iir_df2t in math library. And both crossover.c and crossover_build_blob.m are un-changed, coeffs should be still aligned?

File: tools/topology/m4/crossover.m4
Comment (johnylin76) [reply tools/topology/m4/crossover.m4:29]: After I removed this token, the error messages will appear while "./scripts/build-tools.sh -t"

ALSA lib data.c:700:(build_tuples) cannot find tuples CROSSOVER1.0_crossover_process_tuples_str
ALSA lib parser.c:285:(tplg_build) failed to check topology integrity

It sounds like process_tuples_str is still necessary?

Comment (keyonjie) [reply tools/topology/m4/crossover.m4:29]: if you remove the whole definition of "'N_CROSSOVER($1)`_crossover_process_tuples_str" here, you need to remove the reference to this in line 34 below also.

Comment (johnylin76) [reply tools/topology/m4/crossover.m4:29]: Thanks, that works.
And I guess it still need to define a tuple_uuid type with UUID in crossover.m4 file?
(according to https://github.com/thesofproject/sof/blob/master/tools/tplg_parser/tplg_parser.c#L1395)
Or we should just leave process->comp->type as NONE in tplg_load_process?
(according to https://github.com/thesofproject/sof/blob/master/tools/tplg_parser/tplg_parser.c#L1058)

File: tools/tune/crossover/crossover_build_blob.m
Comment (cujomalainey) [reply tools/tune/crossover/crossover_build_blob.m:1]: you are right, I appear to be making noise again, apparently I had the common sense to parse our config into the struct to catch struct change errors. Can you post a picture of the waveform?

Comment (johnylin76) [reply tools/tune/crossover/crossover_build_blob.m:1]: I found the root cause is that there is a mistake on substituting inline biquad function to math library (iir_df2t) for crossover usage. Please see https://github.com/thesofproject/sof/pull/3263/commits/59cd8942844ad41b46eff9616c99a4a333bfeb85

File: tools/topology/m4/crossover.m4
Comment (johnylin76) [reply tools/topology/m4/crossover.m4:29]: Ah please neglect my last comment because the answer is already in detect.m4 (after I rebased my project)

However, according to https://thesofproject.github.io/latest/developer_guides/uuid/index.html
I think something is still missing (e.g. parse uuid token) on running testbench. I will investigate more.

Comment (johnylin76) [reply tools/topology/m4/crossover.m4:29]: @keyonjie I think the code of parsing uuid token should be ported to tools/tplg_parser from Linux topology driver
(mentioned in https://thesofproject.github.io/latest/developer_guides/uuid/index.html#linux-topology-driver)
if we need to run testbench for topologies with UUID token (and without process_type token).

**Comment by lgirdwood** [comment]
@johnylin76 can you check the build failures. Btw @singalsu is back on Monday.

**Comment by johnylin76** [comment]
> @johnylin76 can you check the build failures. Btw @singalsu is back on Monday.

Thanks. I fixed the issue: crossover.c should use SOF_RT_UUID instead of SOF_UUID

File: tools/topology/m4/crossover.m4
Comment (johnylin76) [inline tools/topology/m4/crossover.m4:17]: FYI I added UUID tuple definition in crossover.m4 just as other components.

File: src/audio/Kconfig
Comment (johnylin76) [inline src/audio/Kconfig:120]: deconfig it, thanks.

File: src/include/ipc/topology.h
Comment (lgirdwood) [reply src/include/ipc/topology.h:50]: @johnylin76 can you delete this change. I think we are good to merge after this.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
We have UUIOD so dont need the comp enum now. Good to go otherwise.

File: src/include/ipc/topology.h
Comment (johnylin76) [reply src/include/ipc/topology.h:50]: Hi @lgirdwood , my only concern is that testbench doesn't seem to have uuid token parsing now, which I think should be ported from Linux topology driver.
(mentioned in https://thesofproject.github.io/latest/developer_guides/uuid/index.html#linux-topology-driver)

Then it will cause unsupported comp type error while running testbench for crossover. Shouldn't we port the uuid token parsing to testbench first?

**Comment by johnylin76** [comment]
Hi @lgirdwood , my only concern is that testbench doesn't seem to have uuid token parsing now, which I think should be ported from Linux topology driver.
(mentioned in https://thesofproject.github.io/latest/developer_guides/uuid/index.html#linux-topology-driver)

Then it will cause unsupported comp type error while running testbench for crossover. Shouldn't we port the uuid token parsing to testbench first?

**Comment by lgirdwood** [comment]
@johnylin76 good point - @keyonjie or @singalsu any update here for support in testbench ?

**Comment by johnylin76** [comment]
Hi, any suggestion?

**Comment by keyonjie** [comment]
> @johnylin76 good point - @keyonjie or @singalsu any update here for support in testbench ?

Yes, we need add this support to testbench, I can try to add it this week.

**Comment by keyonjie** [comment]
> > @johnylin76 good point - @keyonjie or @singalsu any update here for support in testbench ?
> 
> Yes, we need add this support to testbench, I can try to add it this week.

@singalsu @lgirdwood just looking into tplg_parser/topology.h and find that we actually already have significant gap with Linux deriver sof/topology.c, we need to fill up those gaps first.

**Comment by juimonen** [comment]
@keyonjie and others: I find it really bad we have 2 different parsers in linux driver and testbench. There should be some way of detaching the exactly same parsing from the driver to test bench. Now, having tried this way back, it is pretty difficult as the kernel C types don't match user space. But maybe some kind of semi automatic scripting could work?

In essence: if we do for example fuzzing in testbench, it could tell something about the driver's functionality, or it might not tell anything...

**Comment by cujomalainey** [comment]
@bzhg FYI

**Comment by cujomalainey** [comment]
@juimonen I agree, this isn't ideal, it makes fuzzing harder too as we are likely patching irrelevant memory bugs.

**Comment by keyonjie** [comment]
@juimonen exactly. @xiulipan @aiChaoSONG please be noticed that the tplg_parser used in our CI could be out-of-date and not aligned with the Linux driver one. This worth a topic to discuss about how we should proceed.

**Comment by cujomalainey** [comment]
@lgirdwood given the hefty work to get tplg_parser fixed can we land as is for the time being an leave a TODO to remove the enum once tplg_parser is fixed?

File: src/audio/crossover/crossover.c
Comment (keyonjie) [inline src/audio/crossover/crossover.c:38]: Alignment should match open parenthesis.

File: src/include/ipc/topology.h
Comment (keyonjie) [reply src/include/ipc/topology.h:50]: > Hi @lgirdwood , my only concern is that testbench doesn't seem to have uuid token parsing now, which I think should be ported from Linux topology driver.
> (mentioned in https://thesofproject.github.io/latest/developer_guides/uuid/index.html#linux-topology-driver)
> 
> Then it will cause unsupported comp type error while running testbench for crossover. Shouldn't we port the uuid token parsing to testbench first?

@johnylin76 @cujomalainey if nobody is using testbench for crossover ATM, it's fine to remove the adding of SOF_COMP_CROSSOVER and leave the tplg_parser gap as TODO?

**Comment by aiChaoSONG** [comment]
> @juimonen exactly. @xiulipan @aiChaoSONG please be noticed that the tplg_parser used in our CI could be out-of-date and not aligned with the Linux driver one. This worth a topic to discuss about how we should proceed.

Does this PR changes the topology from ABI level,  for example, snd_soc_tplg_damp_widget structure ,etc? if not, then we are good.  Add a new widget type will not influence the topology parser.

**Comment by lgirdwood** [comment]
@juimonen @keyonjie @cujomalainey @johnylin76 I think there are actually 3 parsers.
1) Kernel - GPLv2
2) tplg_parser  - Testebench/Fuzzer - BSD
3) CI - python - ???

tplg_parser already has UUID support APIs, but it does need those APIs glued to component matching code which is currently using comp_type only. It's not a big job, probably a day or two.

@johnylin76 can you remove the new comp_type form this PR and I will merge. I will find someone to fix parser so we can use Cross Over in the pasrer before v1.6.

**Comment by johnylin76** [comment]
Hi @lgirdwood ,
New comp_type SOF_COMP_CROSSOVER and the related change are removed in this PR, thanks.

**Comment by johnylin76** [comment]
> We have UUID so dont need the comp enum now. Good to go otherwise.

Done, thanks.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Jenkins has BSW DUT HW failure.
@johnylin76 testbench and fuzzer UUID support is being worked now by @juimonen

**Comment by cujomalainey** [comment]
@sebcarlucci fyi

**Pull request merged**
